platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/files/DataFileCompactor.java

```diff
@@ -0,0 +1,482 @@
+        currentReader.get().setFileCompleted();
+        currentReader.set(null);
+    }
+
+    /**
+     * Puts file compaction on hold, if it's currently in progress. If not in progress, it will
+     * prevent compaction from starting until {@link #resumeCompaction()} is called. The most
+     * important thing this method does is it makes data files consistent and read only, so they can
+     * be included to snapshots as easily as to create hard links. In particular, if compaction is
+     * in progress, and a new data file is being written to, this file is flushed to disk, no files
+     * are created and no index entries are updated until compaction is resumed.
+     * <p>
+     * This method should not be called on the compaction thread.
+     *
+     * <b>This method must be always balanced with and called before {@link DataFileCompactor#resumeCompaction()}. If
+     * there are more / less calls to resume compactions than to pause, or if they are called in a
+     * wrong order, it will result in deadlocks.</b>
+     *
+     * @throws IOException If an I/O error occurs
+     * @see #resumeCompaction()
+     */
+    public void pauseCompaction() throws IOException {
+        snapshotCompactionLock.acquireUninterruptibly();
+        // Check if compaction is currently in progress. If so, flush and close the current file, so
+        // it's included to the snapshot
+        final DataFileWriter<?> compactionWriter = currentWriter.get();
+        if (compactionWriter != null) {
+            compactionWasInProgress.set(true);
+            compactionLevelInProgress.set(compactionWriter.getMetadata().getCompactionLevel());
+            finishCurrentCompactionFile();
+            // Don't start a new compaction file here, as it would be included to snapshots, but
+            // it shouldn't, as it isn't fully written yet. Instead, a new file will be started
+            // right after snapshot is taken, in resumeCompaction()
+        }
+        // Don't release the lock here, it will be done later in resumeCompaction(). If there is no
+        // compaction currently running, the lock will prevent starting a new one until snapshot is
+        // done
+    }
+
+    /**
+     * Resumes compaction previously put on hold with {@link #pauseCompaction()}. If there was no
+     * compaction running at that moment, but new compaction was started (and blocked) since {@link
+     * #pauseCompaction()}, this new compaction is resumed.
+     *
+     * <b>This method must be always balanced with and called after {@link #pauseCompaction()}. If
+     * there are more / less calls to resume compactions than to pause, or if they are called in a
+     * wrong order, it will result in deadlocks.</b>
+     *
+     * @throws IOException If an I/O error occurs
+     */
+    public void resumeCompaction() throws IOException {
+        try {
+            if (compactionWasInProgress.getAndSet(false)) {
+                assert currentWriter.get() == null;
```

Comment:
The `assert` keyword is only valid in testing.  Do we not want these checks to happen in production?

Reply:
No, it's for testing only. In production it shouldn't be happening at all.